fedex.go

```diff
@@ -46,68 +47,77 @@ func init() {
-		err   error
-	)
+	for delay := 0; delay <= 5; delay++ {
+		pickupOffset := &models.PickupOffset{Days: delay, Hours: 10, Minutes: 45}
+		pickupSuccess := f.bookPickup(pickup, pickupOffset)
+		if pickupSuccess != nil {
+			return pickupSuccess, nil
+		}
+	}
 
 	for delay := 0; delay <= 5; delay++ {
-		fields := log.Fields{"pickup": pickup}
-
-		// Calculate pickup window, but just try the next window in case of error
-		var window *models.PickupTimeWindow
-		window, err = pickupTimeWindow(pickup.PickupLocation.Address, delay)
-		if err != nil {
-			log.WithFields(fields).Error("calculate pickup time", err)
-			continue
+		pickupOffset := &models.PickupOffset{Days: delay, Hours: 10, Minutes: 0}
+		pickupSuccess := f.bookPickup(pickup, pickupOffset)
+		if pickupSuccess != nil {
+			return pickupSuccess, nil
 		}
-		fields["window"] = window
-
-		reply, err = f.API.CreatePickup(pickup, window)
-		switch err.(type) {
-		case nil:
-			fields["reply"] = reply
-			log.WithFields(fields).Info("made pickup")
-			return &models.PickupSuccess{
-				ConfirmationNumber: reply.PickupConfirmationNumber,
-				Window:             *window,
-			}, nil
-
-		case models.PickupAlreadyExistsError:
-			fields["reply"] = reply
-			log.WithFields(fields).Info("pickup already exists")
-			return &models.PickupSuccess{
-				Window: *window,
-			}, nil
-
-		default:
-			fields["err"] = err
-			log.WithFields(fields).Info("failed pickup")
+	}
+
+	return nil, fmt.Errorf("unable to schedule a fedex pickup")
+}
+
+func (f Fedex) bookPickup(pickup *models.Pickup, pickupOffset *models.PickupOffset) *models.PickupSuccess {
+	fields := log.Fields{"pickup": pickup}
+
+	window, err := pickupTimeWindow(clock.NewClock(), pickup.PickupLocation.Address, pickupOffset)
+	if err != nil {
+		log.WithFields(fields).Error("calculate pickup time", err)
+		return nil
+	}
+	fields["window"] = window
+
+	reply, err := f.API.CreatePickup(pickup, window)
+	switch err.(type) {
+	case nil:
+		fields["reply"] = reply
+		log.WithFields(fields).Info("made pickup")
+		return &models.PickupSuccess{
+			ConfirmationNumber: reply.PickupConfirmationNumber,
+			Window:             *window,
 		}
+
+	case models.PickupAlreadyExistsError:
+		fields["reply"] = reply
+		log.WithFields(fields).Info("pickup already exists")
+		return nil
+
+	default:
+		fields["err"] = err
+		log.WithFields(fields).Info("failed pickup")
 	}
 
-	return nil, fmt.Errorf("fedex create pickup: %s", err)
+	return nil
 }
 
-func pickupTimeWindow(pickupAddress models.Address, numDaysToDelay int) (*models.PickupTimeWindow, error) {
+func pickupTimeWindow(clock clock.Clock, pickupAddress models.Address, pickupOffset *models.PickupOffset) (*models.PickupTimeWindow, error) {
 	location, err := toLocation(pickupAddress)
 	if err != nil {
 		location = laTimeZone
 	}
 
-	readyTime := time.Now().In(location).Add(time.Duration(numDaysToDelay*24) * time.Hour)
+	readyTime := time.Now().In(location).Add(time.Duration(pickupOffset.Days*24) * time.Hour)
```

Comment:
Did you mean for this to be clock.Now()?

Reply:
yes, thank you. Updated!